File: cSharp/BallClock/BallClock/BallClock.cs
using Newtonsoft.Json;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace BallClock
{
    class BallClock
    {
        BallQueue queue;
        BallStack hoursStack;
        BallStack fiveMinutesStack;
        BallStack minutesStack;

        public BallClock(int balls)
        {
            queue = new BallQueue(balls);
            hoursStack = new BallStack(11, null, queue);
            fiveMinutesStack = new BallStack(11, hoursStack, queue);
            minutesStack = new BallStack(4, fiveMinutesStack, queue);
        }

        public void Tick()
        {
            queue.Tick(minutesStack);
        }

        public override string ToString()
        {
            string result = "";

            result += "Queue:\n";
            result += queue.ToString();
            result += "\n";
            result += "Minutes:\n";
            result += minutesStack.ToString();
            result += "\n";
            result += "Five Minutes:\n";
            result += fiveMinutesStack.ToString();
            result += "\n";
            result += "Hours:\n";
            result += hoursStack.ToString();
            result += "\n";
            result += "\n";

            return result;
        }

        public bool isStartingOrder()
        {
            int[] balls = queue.ToArray();
            for(int i = 0; i < queue.getCount(); ++i)
            {
                if(balls[i] != i + 1)
                {
                    return false;
                }
            }
            return true;
        }

        public string ToJson()
        {
            var Min = minutesStack.ToArray();
            var FiveMin = fiveMinutesStack.ToArray();
            var Hour = hoursStack.ToArray();
            var Main = queue.ToArray();
            var result = new { Min, FiveMin, Hour, Main };
            return JsonConvert.SerializeObject(result);
        }
    }
}

File: go/ballClock/BallClock/ballClock.go
package BallClock

import (
	"encoding/json"
	"fmt"

	"github.com/cmpickle/ballClock/go/ballClock/collections/queue"
	"github.com/cmpickle/ballClock/go/ballClock/collections/stack"
)

type BallClock struct {
	Count   int
	Min     *stack.Stack `json:"Min"`
	FiveMin *stack.Stack `json:"FiveMin"`
	Hour    *stack.Stack `json:"Hour"`
	Main    *queue.Queue `json:"Main"`
}

type BallClockJson struct {
	Min     []int `json:"Min"`
	FiveMin []int `json:"FiveMin"`
	Hour    []int `json:"Hour"`
	Main    []int `json:"Main"`
}

func New(count int) *BallClock {
	ballClock := &BallClock{count, stack.New(), stack.New(), stack.New(), queue.New()}
	for i := 0; i < count; i++ {
		ballClock.Main.Enqueue(i + 1)
	}
	return ballClock
}

func (ballClock *BallClock) Ticks(minutes int) {
	for i := 0; i < minutes; i++ {
		ballClock.Tick()
	}
}

func (ballClock *BallClock) Tick() {
	ball := ballClock.Main.Dequeue()
	if ball == nil {
		return
	}

	if ballClock.Min.Len() < 4 {
		ballClock.Min.Push(ball)
	} else {
		ballClock.returnBalls(ballClock.Min)

		if ballClock.FiveMin.Len() < 11 {
			ballClock.FiveMin.Push(ball)
		} else {
			ballClock.returnBalls(ballClock.FiveMin)

			if ballClock.Hour.Len() < 11 {
				ballClock.Hour.Push(ball)
			} else {
				ballClock.returnBalls(ballClock.Hour)

				ballClock.Main.Enqueue(ball)
			}
		}
	}
}

func (ballClock *BallClock) returnBalls(stack *stack.Stack) {
	for stack.Len() > 0 {
		ball := stack.Pop()
		if ball != nil {
			ballClock.Main.Enqueue(ball)
		}
	}
}

func (ballClock *BallClock) IsStartingOrder() bool {
	arr := ballClock.Main.ToArray()
	for i := 0; i < ballClock.Main.Len(); i++ {
		if arr[i] != i+1 {
			return false
		}
	}
	return true
}

func (ballClock *BallClock) ToJson() []byte {
	bc := &BallClockJson{Main: ballClock.Main.ToArray(), FiveMin: ballClock.FiveMin.ToArray(), Hour: ballClock.Hour.ToArray(), Min: ballClock.Min.ToArray()}
	result, err := json.Marshal(bc)
	if err != nil {
		fmt.Println("Error parsing Ball Clock to Json")
	}
	return result
}

func (ballClock *BallClock) ToString() string {
	var result string
	for i := 0; i < ballClock.Main.Len(); i++ {
		result += " " + fmt.Sprintf("%v", ballClock.Main.ElementAt(i))
	}
	return result
}
